modules/apps/commerce/commerce-wish-list-service/src/main/java/com/liferay/commerce/wish/list/service/impl/CommerceWishListItemServiceImpl.java

```diff
@@ -52,8 +51,26 @@ public CommerceWishListItem addCommerceWishListItem(
 			cProduct.getPublishedCPDefinitionId());
 
 		return commerceWishListItemLocalService.addCommerceWishListItem(
-			commerceWishListId, cProductId, cpInstanceUuid, json,
-			serviceContext);
+			getUserId(), commerceWishListId, cpInstanceUuid, cProductId, json);
+	}
+
+	@Override
+	public CommerceWishListItem addOrUpdateCommerceWishListItem(
+			long commerceAccountId, long commerceWishListId,
+			String cpInstanceUuid, long cProductId, String json)
+		throws PortalException {
+
+		_commerceWishListModelResourcePermission.check(
+			getPermissionChecker(), commerceWishListId, ActionKeys.UPDATE);
+
+		CProduct cProduct = cProductLocalService.getCProduct(cProductId);
+
+		commerceProductViewPermission.check(
+			getPermissionChecker(), commerceAccountId,
+			cProduct.getPublishedCPDefinitionId());
```

Comment:
something to evaluate: do before the check of the product visibility and then the wishList permission